Broken on node 20.12.x

**Describe the bug**
net-keepalive fails to install on node 20.12.x onwards.

**To Reproduce**
```
$ nvm use 20.12
Now using node v20.12.1 (npm v10.5.0)
$ npm i net-keepalive
...
npm ERR! /home/otavio/.cache/node-gyp/20.12.1/include/node/js_native_api.h:523:43: note:   initializing argument 4 of ‘napi_status napi_add_finalizer(napi_env, napi_value, void*, node_api_nogc_finalize, void*, napi_ref__**)’
npm ERR!   523 |                    node_api_nogc_finalize finalize_cb,
npm ERR!       |                    ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
npm ERR! make: *** [ffi_bindings.target.mk:121: Release/obj.target/ffi_bindings/src/ffi.o] Error 1
npm ERR! gyp ERR! build error 
npm ERR! gyp ERR! stack Error: `make` failed with exit code: 2
npm ERR! gyp ERR! stack at ChildProcess.<anonymous> (/home/otavio/.nvm/versions/node/v20.12.1/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:209:23)
npm ERR! gyp ERR! System Linux 6.5.0-26-generic
npm ERR! gyp ERR! command "/home/otavio/.nvm/versions/node/v20.12.1/bin/node" "/home/otavio/.nvm/versions/node/v20.12.1/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd /home/otavio/git/test-napi/node_modules/ffi-napi
npm ERR! gyp ERR! node -v v20.12.1
npm ERR! gyp ERR! node-gyp -v v10.0.1
npm ERR! gyp ERR! not ok

```

**Expected behavior**
net-keepalive would install :)

**Package (please complete the following information):**
 - OS: all
 - Node: 20.12.x
 - Version: 3.0.1

**Additional context**
This is failing due to a breaking change on node please see https://github.com/nodejs/node/issues/52229
I think also ff-napi sadly is unmantained: https://github.com/node-ffi-napi/node-ffi-napi/issues/269
Do you have any plans or are considering migrating to any of the possible substitutes?

--- a/.github/workflows/pull-request.yaml
+++ b/.github/workflows/pull-request.yaml
@@ -72,7 +72,6 @@ jobs:
           - windows-latest
         node: # https://nodejs.org/en/about/releases/
           - '*'
-          - '22'
           - '20'
           - '18'
 

--- a/.github/workflows/pull-request.yaml
+++ b/.github/workflows/pull-request.yaml
@@ -71,8 +71,10 @@ jobs:
           - macos-latest
           - windows-latest
         node: # https://nodejs.org/en/about/releases/
-          - '*' # Current
-          - '18' # Active LTS
+          - '*'
+          - '22'
+          - '20'
+          - '18'
 
       fail-fast: false
 

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -15,8 +15,10 @@ jobs:
           - macos-latest
           - windows-latest
         node: # https://nodejs.org/en/about/releases/
-          - '*' # Current
-          - '18' # Active LTS
+          - '*'
+          - '22'
+          - '20'
+          - '18'
       fail-fast: false
 
     steps:

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -16,7 +16,6 @@ jobs:
           - windows-latest
         node: # https://nodejs.org/en/about/releases/
           - '*'
-          - '22'
           - '20'
           - '18'
       fail-fast: false

--- a/lib/ffi-bindings.js
+++ b/lib/ffi-bindings.js
@@ -1,26 +1,45 @@
 const { platform } = require('os')
 const { errnoException } = require('./commons')
-const Ref = require('ref-napi')
-const FFI = require('ffi-napi')
+const { load, DataType, open, close, arrayConstructor } = require('ffi-rs')
+
+const LIBRARY_NAME = 'libnative'
 
 const createFFI = () => {
-  const cInt = Ref.types.int
-  const cVoid = Ref.types.void
+  const cInt = DataType.I32
+  const cVoidRef = DataType.External
 
-  return FFI.Library(null, {
-    //name       ret    1     2     3     4                   5
-    setsockopt: [cInt, [cInt, cInt, cInt, Ref.refType(cVoid), cInt]],
-    getsockopt: [
-      cInt,
-      [cInt, cInt, cInt, Ref.refType(cVoid), Ref.refType(cInt)],
-    ],
-  })
+  return {
+    setsockopt: (fd, level, name, value, valueLength) => {
+      const ret = load({
+        library: LIBRARY_NAME,
+        funcName: 'setsockopt',
+        retType: cInt,
+        paramsType: [cInt, cInt, cInt, cVoidRef, cInt],
+        paramsValue: [fd, level, name, value, valueLength],
+      })
+      return ret
+    },
+    getsockopt: (fd, level, name, value, valueLength) => {
+      const ret = load({
+        library: LIBRARY_NAME,
+        funcName: 'getsockopt',
+        retType: cInt,
+        paramsType: [cInt, cInt, cInt, cVoidRef, cVoidRef],
+        paramsValue: [fd, level, name, value, valueLength],
+      })
+      return ret
+    },
+  }
 }
 
 const ffi = (() => {
   let instance
   return () => {
     if (!instance) {
+      open({
+        library: LIBRARY_NAME,
+        path: '',
+      })
       instance = createFFI()
     }
     return instance
@@ -28,14 +47,14 @@ const ffi = (() => {
 })()
 
 const setsockopt = (fd, level, name, value, valueLength) => {
-  if(fd == null) {
+  if (fd == null) {
     return false
   }
 
   const err = ffi().setsockopt(fd, level, name, value, valueLength)
 
   if (err !== 0) {
-    const errno = FFI.errno()
+    const errno = 9 // FIXME: there's no FFI.errno() in ffi-rs
     throw errnoException(errno, 'setsockopt')
   }
 
@@ -43,14 +62,14 @@ const setsockopt = (fd, level, name, value, valueLength) => {
 }
 
 const getsockopt = (fd, level, name, value, valueLength) => {
-  if(fd == null) {
+  if (fd == null) {
     return false
   }
 
   const err = ffi().getsockopt(fd, level, name, value, valueLength)
 
   if (err !== 0) {
-    const errno = FFI.errno()
+    const errno = 9 // FIXME: there's no FFI.errno() in ffi-rs
     throw errnoException(errno, 'getsockopt')
   }
   return true

--- a/lib/index.js
+++ b/lib/index.js
@@ -180,8 +180,12 @@ module.exports.setKeepAliveProbes = function setKeepAliveProbes(socket, cnt) {
 
   const fd = _getSocketFD(socket),
     count = ~~cnt,
-    cntVal = Ref.alloc('int', count),
-    cntValLn = cntVal.type.size
+    dataType = DataType.I32,
+    [cntVal] = createPointer({
+      paramsType: [dataType],
+      paramsValue: [count],
+    }),
+    cntValLn = 4 // sizeof int
 
   return FFIBindings.setsockopt(
     fd,
@@ -219,8 +223,15 @@ module.exports.getKeepAliveProbes = function getKeepAliveProbes(socket) {
   )
 
   const fd = _getSocketFD(socket),
-    cntVal = Ref.alloc(Ref.types.int),
-    cntValLn = Ref.alloc(Ref.types.int, cntVal.type.size)
+    dataType = DataType.I32,
+    [cntVal] = createPointer({
+      paramsType: [dataType],
+      paramsValue: [0],
+    }),
+    [cntValLn] = createPointer({
+      paramsType: [DataType.I32],
+      paramsValue: [4], // sizeof int
+    })
 
   FFIBindings.getsockopt(
     fd,
@@ -230,7 +241,12 @@ module.exports.getKeepAliveProbes = function getKeepAliveProbes(socket) {
     cntValLn
   )
 
-  return cntVal.deref()
+  const [value] = restorePointer({
+    retType: [dataType],
+    paramsValue: [cntVal],
+  })
+
+  return value
 }
 
 /**

--- a/lib/index.js
+++ b/lib/index.js
@@ -296,8 +296,12 @@ module.exports.setUserTimeout = function setUserTimeout(socket, msecs) {
 
   const fd = _getSocketFD(socket),
     msecInt = ~~msecs,
-    msecVal = Ref.alloc('int', msecInt),
-    msecValLn = msecVal.type.size
+    dataType = DataType.I32,
+    [msecVal] = createPointer({
+      paramsType: [dataType],
+      paramsValue: [msecInt],
+    }),
+    msecValLn = 4 // sizeof int
 
   return FFIBindings.setsockopt(
     fd,
@@ -342,8 +346,15 @@ module.exports.getUserTimeout = function getUserTimeout(socket) {
   )
 
   const fd = _getSocketFD(socket),
-    msecVal = Ref.alloc(Ref.types.uint32),
-    msecValLn = Ref.alloc(Ref.types.uint32, msecVal.type.size)
+    dataType = DataType.I32,
+    [msecVal] = createPointer({
+      paramsType: [dataType],
+      paramsValue: [0],
+    }),
+    [msecValLn] = createPointer({
+      paramsType: [DataType.I32],
+      paramsValue: [4], // sizeof int
+    })
 
   FFIBindings.getsockopt(
     fd,
@@ -353,5 +364,10 @@ module.exports.getUserTimeout = function getUserTimeout(socket) {
     msecValLn
   )
 
-  return msecVal.deref()
+  const [value] = restorePointer({
+    retType: [dataType],
+    paramsValue: [msecVal],
+  })
+
+  return value
 }

--- a/lib/index.js
+++ b/lib/index.js
@@ -1,6 +1,6 @@
 /**
  * The Missing (TCP_KEEPINTVL and TCP_KEEPCNT) SO_KEEPALIVE socket option setters and getters for Node using
- * ffi-napi module.
+ * ffi-rs module.
  *
  * Note: For methods provided by this module to work you must enable SO_KEEPALIVE and set the TCP_KEEPIDLE options for
  * socket using Net.Socket-s built in method socket.setKeepAlive([enable][, initialDelay]) !
@@ -22,7 +22,7 @@ const Assert = require('assert'),
   Net = require('net'),
   OS = require('os'),
   Constants = require('./constants'),
-  Ref = require('ref-napi'),
+  { DataType, createPointer, restorePointer } = require('ffi-rs'),
   FFIBindings = require('./ffi-bindings')
 
 const _isSocket = (socket) => socket instanceof Net.Socket
@@ -75,8 +75,12 @@ module.exports.setKeepAliveInterval = function setKeepAliveInterval(
 
   const fd = _getSocketFD(socket),
     seconds = ~~(msecs / 1000),
-    intvlVal = Ref.alloc('int', seconds),
-    intvlValLn = intvlVal.type.size
+    dataType = DataType.I32,
+    [intvlVal] = createPointer({
+      paramsType: [dataType],
+      paramsValue: [seconds],
+    }),
+    intvlValLn = 4 // sizeof int
 
   return FFIBindings.setsockopt(
     fd,
@@ -114,8 +118,15 @@ module.exports.getKeepAliveInterval = function getKeepAliveInterval(socket) {
   )
 
   const fd = _getSocketFD(socket),
-    intvlVal = Ref.alloc(Ref.types.uint32),
-    intvlValLn = Ref.alloc(Ref.types.uint32, intvlVal.type.size)
+    dataType = DataType.I32,
+    [intvlVal] = createPointer({
+      paramsType: [dataType],
+      paramsValue: [0],
+    }),
+    [intvlValLn] = createPointer({
+      paramsType: [DataType.I32],
+      paramsValue: [4], // sizeof int
+    })
 
   FFIBindings.getsockopt(
     fd,
@@ -125,7 +136,12 @@ module.exports.getKeepAliveInterval = function getKeepAliveInterval(socket) {
     intvlValLn
   )
 
-  return intvlVal.deref() * 1000
+  const [value] = restorePointer({
+    retType: [dataType],
+    paramsValue: [intvlVal],
+  })
+
+  return value * 1000
 }
 
 /**

--- a/package.json
+++ b/package.json
@@ -50,8 +50,7 @@
   },
   "homepage": "https://hertzg.github.io/node-net-keepalive/",
   "dependencies": {
-    "ffi-napi": "^4.0.1",
-    "ref-napi": "^3.0.0"
+    "ffi-rs": "^1.0.58"
   },
   "devDependencies": {
     "@commitlint/cli": "^18.0.0",

--- a/package.json
+++ b/package.json
@@ -50,7 +50,7 @@
   },
   "homepage": "https://hertzg.github.io/node-net-keepalive/",
   "dependencies": {
-    "ffi-rs": "^1.0.58"
+    "ffi-rs": "1.0.58"
   },
   "devDependencies": {
     "@commitlint/cli": "^18.0.0",

